lib/interfaces.ts

```diff
@@ -0,0 +1,77 @@
+import { ComputedRef, DeepReadonly, Ref, UnwrapNestedRefs, WatchCallback } from 'vue'
+
+export interface IModule<
+  S extends IState,
+  G extends IGetters,
+  M extends IMutations,
+  SM extends ISubModules = never,
+  P extends IPlugins<S> = IPlugins<S>,
+> {
+  readonly options: IModuleOptions<S, G, M, SM, P>
+  flatten(): IFlattenedModule<S, G, M, SM>
+}
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+export type ISubModules = Record<string, IModule<any, any, any, any, any>>
+
+export type IState = Record<string, any>
+
+export type IGetters = Record<string, () => any>
+
+export type IMutations = Record<string, (...args: any) => void>
+
+export type IPlugins<S extends IState> = [
+  ...plugins: Array<IPlugin<S>>,
+  lastPlugin: ILastPlugin<S>,
+]
+
+export interface IPlugin<S extends IState> {
+  onStateInit?: (state: Partial<S>) => Partial<S>
+  onDataChange?: WatchCallback<UnwrapNestedRefs<S>, UnwrapNestedRefs<S> | undefined>
+}
+
+export interface ILastPlugin<S extends IState> extends Omit<IPlugin<S>, 'onStateInit'> {
+  onStateInit?: (state: Partial<S>) => S
+}
```

Comment:
So as we discussed before, we have to restrict the output state to be the same type as the input state.
We allowed developers to use as many preceeding plugins as they want, as long as the last plugin is returning something with the same type as the input state, everything should work as expected.
But that make me have to split the interface for plugin into 2 version—one returns `Partial<T>`, the other returns `<T>`.
You might have guessed it already, that'll cause some problem when applying plugins into module.
I'll describe it in detail in `example/src/modules/user.ts`.

Reply:
Updated in latest commit.